### PR TITLE
feat: Support binding text to container via context menu

### PR DESCRIPTION
**Pull request by ad1992**
https://user-images.githubusercontent.com/11256141/158994766-917d6fe8-5bc4-40ea-90e5-8dbf17c23ffa.mp4

- [x] Instead of binding to closest container, bind to selected container (container and text selected together)
- [x] Allow binding only if container doesn't have bound text
- [x] Text shouldn't be selected once binded

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -5449,6 +5467,7 @@ class App extends React.Component<AppProps, AppState> {
             ((probablySupportsClipboardBlob && elements.length > 0) ||
               (probablySupportsClipboardWriteText && elements.length > 0)) &&
               separator,
+            allowBinding && actionBindText,
```
Comment (dwelle) [inline src/components/App.tsx:5470]: we shouldn't show this when clicking on canvas as we show element-specific actions only when clicking on elements

Comment (ad1992) [reply src/components/App.tsx:5470]: But in this case it does make sense to show it when clicking on canvas too coz its related to both the elements and not just a single one ? similar to copy styles as png/svg. Also I am not sure why we don't show the actions when elements selected and clicked on canvas coz when user is selecting elements they want to use the actions on those elements so it makes more intutive to show those actions even when clicked on canvas

File: src/components/App.tsx [hunk 2]
```diff
@@ -5404,6 +5405,23 @@ class App extends React.Component<AppProps, AppState> {
...
+        isTextElement(selectedElements[0]) ||
+        isTextElement(selectedElements[1]);
+
+      const bindingContainer =
+        isTextBindableContainer(selectedElements[0]) ||
+        isTextBindableContainer(selectedElements[1]);
+      if (textElement && bindingContainer) {
+        allowBinding = true;
+      }
+    }
```
Comment (dwelle) [inline src/components/App.tsx:5423]: 1) let's put this into `action.contextMenuPredicate` instead
2) as discussed on chat, if we're not gonna allow replacing for now, then we should not show the item if the container already has a bound text

File: src/actions/actionBoundText.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    redrawTextBoundingBox(textElement, container);
+    const updatedElements = elements.slice();
+    const textElementIndex = updatedElements.findIndex(
+      (ele) => ele.id === textElement.id,
+    );
+    updatedElements.splice(textElementIndex, 1);
+    const containerIndex = updatedElements.findIndex(
+      (ele) => ele.id === container!.id,
+    );
+    updatedElements.splice(containerIndex + 1, 0, textElement!);
```
Comment (dwelle) [inline src/actions/actionBoundText.tsx:96]:
```suggestion
    updatedElements.splice(containerIndex + 1, 0, textElement);
```

File: src/actions/actionBoundText.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      isTextElement(selectedElements[0]) &&
+      isTextBindableContainer(selectedElements[1])
+    ) {
+      textElement = selectedElements[0];
+      container = selectedElements[1];
+    } else {
+      textElement = selectedElements[1] as ExcalidrawTextElement;
+      container = selectedElements[0] as ExcalidrawTextContainer;
+    }
+    mutateElement(textElement!, {
```
Comment (dwelle) [inline src/actions/actionBoundText.tsx:77]:
```suggestion
    mutateElement(textElement, {
```

File: src/components/App.tsx (hunk 1 above)
Comment (dwelle) [reply src/components/App.tsx:5477]: > Also I am not sure why we don't show the actions when elements selected and clicked on canvas coz when user is selecting elements they want to use the actions on those elements so it makes more intutive to show those actions even when clicked on canvas

That cannot be the assumption. The standard is that clicking on canvas means you want canvas actions. Some apps even auto-deselect when click on canvas, and we _may_ wanna do so later as well.

Either way, the way Excalidraw currently works is that clicking on canvas _does not_ show element actions, no matter of selection, so we shouldn't break that assumption just for this action.

Comment (dwelle) [reply src/components/App.tsx:5477]: (note that "element" in this context means the common bounding box)

Comment (ad1992) [reply src/components/App.tsx]: > (note that "element" in this context means the common bounding box)
No we show canvas actions for common bounding box as well thats why I said its confusing and we should element actions when selection when clicking within selection

Comment (dwelle) [reply src/components/App.tsx]: > No we show canvas actions for common bounding box as well thats why I said its confusing and we should element actions when selection when clicking within selection

Huh. Is that a recent regression or was I mistaken? Well, let's fix that in another PR.

**Review by dwelle — APPROVED** [review]

File: src/components/App.tsx (hunk 1 above)
Comment (ad1992) [reply src/components/App.tsx]: Not sure, will check, I just noticed it now and hence started the discussion

**Pull request merged**